### PR TITLE
#8970 Minion tab in Pinot UI

### PR DESCRIPTION
**Pull request by satishwaghela**
Closes: https://github.com/apache/pinot/issues/8970

### CODE REVIEW

**Comment by npawar** [comment]
Thanks @satishwaghela ! Just played around with this and it's pretty neat! Here's some suggestions:

1. On Homepage, Move MINIONS after SERVERS (Order of these tables should be Controllers, Brokers, Servers, Minions)
Screen Shot 2022-07-03 at 12 15 41 AM

2. Is this validation coming from UI or backend? If UI, no need for this restriction in tags for Minions, it can be anything.
Screen Shot 2022-07-03 at 12 18 12 AM

3. After editing the tag for Minion, I see this table appear below, which should not be there
Screen Shot 2022-07-03 at 12 18 31 AM

4. Replace Execution Time with “Elapsed Time”. Calculate the value using System.currentTimeMillis() - startTime. Show value in minutes if < 60. In hours + minutes if > 60
Screen Shot 2022-07-03 at 12 27 12 AM

5. After clicking Schedule Now, show a modal with message “Are you sure you want to schedule this task?” If user clicks YES, then show the Task ID from the response as “<Task id> scheduled” and give user an option to dismiss the message
Screen Shot 2022-07-03 at 1 02 11 AM

**Comment by npawar** [comment]
1. These buttons have different tooltip text provided in the design doc
2. Show every unique table name only once (in this case githubEvents_REALTIME should appear only once)
3. On this page, when clicking “Resume All” show similar modal to the one as used in “Stop All”. Same for Cleanup All.
4. After clicking “Stop All” immediately call the API to update “Task Queue Status” field (it should become STOPPED as per the API). Same for “Resume All”, call API to get updated status
5. After calling Stop or Resume, sometimes going back to homepage just keeps spinning
Screen Shot 2022-07-03 at 12 32 42 AM

**Comment by npawar** [comment]
@joshigaurava @shahsank3t please help review from code perspective. Thanks!

**Comment by npawar** [comment]
Thanks @satishwaghela . I have last couple of comments

1. After scheduling task using “Schedule Now”, show different message as per response. If response contains null, show message “Could not schedule task for <task type>”. If response contains a task id, show the Task ID from the response as “<Task id> scheduled” and give user an option to dismiss the message (dont auto disappear)
2. From this page (http://localhost:9000/#/task-queue/MergeRollupTask/tables/githubEvents_OFFLINE) if i click refresh on browser, I see it takes 10-15 seconds to load.
3. The Start Time is empty for on UI for a task with status IN_PROGRESS, but if I inspect the response, I do see the startTime field.
Screen Shot 2022-07-05 at 6 42 04 PM

File: pinot-controller/src/main/resources/app/utils/PinotMethodUtils.ts [hunk 1]
```diff
@@ -746,6 +761,113 @@ const deleteInstance = (instanceName) => {
...
+};
+
+const getAllTaskTypes = async () => {
+  const finalResponse = {
+    columns: ['Task Type', 'Num Tasks in Queue', 'Queue Status'],
+    records: []
+  }
+  await new Promise((resolve, reject) => {
+    getTaskTypes().then(async (response)=>{
+      if (_.isArray(response.data)) {
```
Comment (joshigaurava) [inline pinot-controller/src/main/resources/app/utils/PinotMethodUtils.ts:777]: Let's fix lodash imports in this file. Importing the entire library is not a good idea. Let's change it to `import { a, b, c } from 'lodash';`

File: pinot-controller/src/main/resources/app/pages/TaskDetail.tsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+          <Grid item xs={12}>
+            <strong>Finish Time:</strong> {get(taskDebugData, 'subtaskInfos.0.finishTime', '')}
+          </Grid>
+          <Grid item xs={12}>
+            <strong>Number of Sub Tasks:</strong> {get(taskDebugData, 'subtaskCount.total', '')}
+          </Grid>
+        </Grid>
+      </div>
+      <Grid container spacing={2}>
+        {/* <Grid item xs={6}>
```
Comment (joshigaurava) [inline pinot-controller/src/main/resources/app/pages/TaskDetail.tsx:130]: If the commented out code is not required, let's remove it before merging.

File: pinot-controller/src/main/resources/app/pages/InstanceListingPage.tsx [hunk 3]
```diff
@@ -63,6 +70,7 @@ const InstanceListingPage = () => {
   ) : (
     <Grid item xs className={classes.gridContainer}>
       <Instances instances={instances} clusterName={clusterName} />
+      {/* {isController && periodicTasks.content} */}
```
Comment (joshigaurava) [inline pinot-controller/src/main/resources/app/pages/InstanceListingPage.tsx:73]: Is this going to be updated?

File: pinot-controller/src/main/resources/app/pages/InstanceDetails.tsx [hunk 4]
```diff
@@ -308,7 +315,7 @@ const InstanceDetails = ({ match }: RouteComponentProps<Props>) => {
               </CustomButton>
               <CustomButton
                 onClick={handleDropAction}
-                tooltipTitle="Removes the node from the cluster. Untag & rebalance (to ensure node is not being used by any table), and shutdown instance, before dropping."
+                tooltipTitle={!instanceName.startsWith('Minion_') ? "Removes the node from the cluster. Untag & rebalance (to ensure node is not being used by any table), and shutdown instance, before dropping." : ""}
```
Comment (joshigaurava) [inline pinot-controller/src/main/resources/app/pages/InstanceDetails.tsx:318]: Removes the node from the cluster. Untag and rebalance (to ensure the node is not being used by any table) and shutdown the instance before dropping.

@npawar

Comment (joshigaurava) [reply pinot-controller/src/main/resources/app/pages/InstanceDetails.tsx:318]: Is there a difference between line 74

```
else if (instanceName.toLowerCase().startsWith('minion')) {
    instanceType = 'MINION';
  }
```

and the condition here? Could we just use `instanceType === 'MINION'` here?

File: pinot-controller/src/main/resources/app/pages/InstanceDetails.tsx [hunk 5]
```diff
@@ -68,9 +68,16 @@ type Props = {
 const InstanceDetails = ({ match }: RouteComponentProps<Props>) => {
   const classes = useStyles();
   const {instanceName} = match.params;
+  let instanceType;
+  if (instanceName.toLowerCase().startsWith('broker')) {
+    instanceType = 'BROKER';
```
Comment (joshigaurava) [inline pinot-controller/src/main/resources/app/pages/InstanceDetails.tsx:73]: Is it feasible to declare these as enums/constants?

**Comment by npawar** [comment]
Still seeing high load time when I refresh this page. Seems to be coming from "info" call?
Screen Shot 2022-07-07 at 11 49 40 AM

**Comment by npawar** [comment]
> also, please take a look at why the Pinot Integration Test fails in the Github Actions. Seems to be throwing some exceptions from ui code

Seeing this 
```
 ERROR in /home/runner/work/pinot/pinot/pinot-controller/src/main/resources/app/pages/TaskQueueTable.tsx
[INFO] ./app/pages/TaskQueueTable.tsx
[INFO] [tsl] ERROR in /home/runner/work/pinot/pinot/pinot-controller/src/main/resources/app/pages/TaskQueueTable.tsx(31,31)
[INFO]       TS2307: Cannot find module '../components/useMinionMetadata'.
[INFO] Child HtmlWebpackCompiler:
[INFO]      1 asset
[INFO]     Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
[INFO]     [0] ./node_modules/html-webpack-plugin/lib/loader.js!./app/index.html 1.43 KiB {0} [built]
[INFO] npm ERR! code ELIFECYCLE
[INFO] npm ERR! errno 2
[INFO] npm ERR! pinot-controller-ui@1.0.0 build: `webpack --mode production`
[INFO] npm ERR! Exit status 2
[INFO] npm ERR! 
[INFO] npm ERR! Failed at the pinot-controller-ui@1.0.0 build script.
[INFO] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**Review by joshigaurava — APPROVED** [review]

**Comment by joshigaurava** [comment]
Looks good!

**Comment by npawar** [comment]
@klsince i'll merge this for now. if you do have any feedback after you play with it, we can address those separately

**Pull request merged**